src/quantumchannel.jl

```diff
@@ -0,0 +1,89 @@
+"""
+QuantumChannel for transmitting qubits as `RegRef`s from one register to another in a quantum protocol simulation, 
+with a channel delay and under the influence of a background process.
+The function `put!` is used to put the `RegRef` containing the qubit in the channel, which can then be received by
+the receiving register after a specified delay using the take! method in a synchronous way.
+
+```jldoctest
+julia> using QuantumSavory; using ResumableFunctions; using ConcurrentSim
+
+julia> bell = (Z1⊗Z1 + Z2⊗Z2)/sqrt(2.0); regA = Register(2); regB = Register(2); initialize!((regA[1], regB[2]), bell);
+
+julia> sim = Simulation();
+
+julia> queue = DelayQueue{RegRef}(sim, 10.0)
+DelayQueue{RegRef}(QueueStore{RegRef, Int64}, 10.0)
+
+julia> qc = QuantumChannel(queue)
+QuantumChannel(Qubit(), DelayQueue{RegRef}(QueueStore{RegRef, Int64}, 10.0), nothing)
+
+julia> @resumable function alice_node(env, qc)
+            println("Putting Alice's qubit in the channel at ", now(env))
+            put!(qc, regA[1])
+        end
+alice_node (generic function with 1 method)
+
+julia> @resumable function bob_node(env, qc)
+            @yield @process take!(env, qc, regB[1]) # wait for the process with delay to complete
+            println("Taking the qubit from alice at ", now(env))
+        end
+bob_node (generic function with 1 method)
+
+julia> @process alice_node(sim, qc); @process bob_node(sim, qc);
+
+julia> run(sim)
+Putting Alice's qubit in the channel at 0.0
+Taking the qubit from alice at 10.0
+
+julia> regA
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    nothing
+    nothing
+
+julia> regB
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    Subsystem 1 of QuantumOpticsBase.Ket 12382959472027850978
+    Subsystem 2 of QuantumOpticsBase.Ket 12382959472027850978
+```
+"""
+struct QuantumChannel
+    trait::Qubit
+    queue::ConcurrentSim.DelayQueue{RegRef}
+    background::Any
+end
+
+function QuantumChannel(queue::ConcurrentSim.DelayQueue{RegRef}, background=nothing)
+    QuantumChannel(Qubit(), queue, background)
+end
+
+function Base.put!(qc::QuantumChannel, rref::RegRef, Δt=nothing)
+    if xor(isnothing(qc.background), isnothing(Δt))
+        throw(ArgumentError(lazy"""
+        Either both background and Δt should be nothing or both should be initialized to appropriate values
+        """))
+    elseif !isnothing(Δt) # if both are not nothing
+        uptotime!(rref.reg.staterefs[rref.idx], rref.reg.stateindices[rref.idx], qc.background, Δt)
+    end
+
+    put!(qc.queue, rref)
```

Comment:
The `rref` should be "emptied" when the `put!` happens, not when the `take!` resolves, because the user might want to immediately reuse the `rref` for something else.

The cleanest way to do that right now is probably on each put! to create a temporary register:

```
uptotime!(rref, now(qc.queue.environment))
channel_reg = Register(1, [qc.background], [qc.trait])
swap!(rref, channel_reg[1])
uptotime!(channel_reg[1], qc.queue.delay)
put!(qc.queue, channel_reg)
```

then in the `take!` you should have

```
uptotime!(channel_reg[1], now(qc.queue.environment))
swap!(...)
```